node/router/src/handshake.rs

```
@@ -134,7 +134,7 @@ impl<N: Network> Router<N> {
         if let Some(addr) = listener_addr {
             if let Ok(ref cr) = handshake_result {
                 if let Some(peer) = self.peer_pool.write().get_mut(&addr) {
-                    self.resolver.write().insert_peer(peer.listener_addr(), peer_addr);
+                    self.resolver.write().insert_peer(peer.listener_addr(), peer_addr, cr.address);
                     peer.upgrade_to_connected(peer_addr, cr.listener_port, cr.address, cr.node_type, cr.version);
                 }
```

Unrelated to your changes, but I recently loooked at this section as well: Is it not a problem when the peer is not in the peer pool? Does it make sense to silently fail here, or should we at least log?

---

You mean the `upgrade_to_connected`? It's impossible for the peer to not belong in the peer pool at this point, as it's right underneath `if let Some(peer) = self.peer_pool.write().get_mut(&addr)`.

---

I mean for the case when `self.peer_pool.write().get_mut(&addr)` returns `None` it will still log   `Completed the handshake with '{peer_addr}'`. Does that make sense?

---

Marking this as resolved, though, as this is something holding up merging the PR.

---

ah, the bit underneath this snippet; it's impossible for a peer to not be in the peer pool at this point, as the handshake has succeeded, and the related `Peer` is in `Connecting` state, meaning any concurrent handshake attempts can't change that state until it's changed by this function; the log and the metrics update are put outside of the lifetime of the guard to reduce its grip on the peer pool